automata/fa/fa.py

```diff
@@ -2,11 +2,245 @@
+            arrow_size (float, optional): Arrow head size. Defaults to 0.85.
+            state_separation (float, optional): Node distance. Defaults to 0.5.
+        Returns:
+            Digraph: The graph in dot format.
+        """
+        # Converting to graphviz preferred input type,
+        # keeping the conventional input styles; i.e fig_size(8,8)
+        # TODO why is (8, 8) the "conventional" size?
+        fig_size = ", ".join(map(str, fig_size))
+        font_size = str(font_size)
+        arrow_size = str(arrow_size)
+        state_separation = str(state_separation)
+
+        # Defining the graph.
+        graph = graphviz.Digraph(strict=False, engine=engine)
+        graph.attr(
+            # size=fig_size, # TODO maybe needed?
+            ranksep=state_separation,
+        )
+        if horizontal:
+            graph.attr(rankdir="LR")
+        if reverse_orientation:
+            if horizontal:
+                graph.attr(rankdir="RL")
+            else:
+                graph.attr(rankdir="BT")
+
+        for state in self.iter_states():
+            # every edge needs an origin node, so we add a null node for every
+            # initial state.
+            if self.is_initial(state):
+                # we use a random uuid to make sure that the null node has a
+                # unique id to avoid colliding with other states and null_nodes.
+                null_node = str(uuid.uuid4())
+                graph.node(
+                    null_node,
+                    label="",
+                    shape='point',
+                    fontsize=font_size,
+                )
+                node = self.get_state_name(state)
+                edge_label = "->" + node
+                graph.edge(
+                    null_node,
+                    node,
+                    tooltip=edge_label,
+                    arrowsize=arrow_size,
+                )
+
+        for state in self.iter_states():
+            shape = 'doublecircle' if self.is_accepting(state) else 'circle'
+            node = self.get_state_name(state)
+            graph.node(node, shape=shape, fontsize=font_size)
+
+        if input_str is None:
+            edge_labels = defaultdict(list)
+            for from_state, to_state, symbol in self.iter_transitions():
+                # TODO only do this for NFA
+                label = "ε" if symbol == "" else str(symbol)
+                from_node = self.get_state_name(from_state)
+                to_node = self.get_state_name(to_state)
+                edge_labels[from_node, to_node].append(label)
+
+            for (from_node, to_node), labels in edge_labels.items():
+                label = ",".join(sorted(labels))
+                graph.edge(
+                    from_node,
+                    to_node,
+                    label=label,
+                    arrowsize=arrow_size,
+                    fontsize=font_size,
+                )
+        else:
+            # TODO
```

Comment:
Maybe the way to do this is to just call `read_input_stepwise` from the parent class? I know this won't work for the GNFA but I don't think it makes sense to support an input string there. It would be cool to visualize the steps in the GNFA -> regex algorithm though.

Reply:
@eliotwrobson Well, `GNFA.read_input_stepwise` already raises the appropriate `NotImplementedError`, so I think it would be fine to call `read_input_stepwise` here. cc @khoda81